gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/GetProjectInvocations.java

```diff
@@ -0,0 +1,134 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.guide.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.util.TreeScanner;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Type.ClassType;
+import java.util.List;
+
+@AutoService(BugChecker.class)
+@BugPattern(severity = SeverityLevel.ERROR, summary = GetProjectInvocations.SUMMARY)
+public final class GetProjectInvocations extends GradleGuideBugChecker implements BugChecker.MethodTreeMatcher {
+    private static final Supplier<ClassSymbol> ACTION_SYM =
+            VisitorState.memoize(s -> (ClassSymbol) s.getSymbolFromString("org.gradle.api.Action"));
+    private static final Supplier<ClassSymbol> TASK_SYM =
+            VisitorState.memoize(s -> (ClassSymbol) s.getSymbolFromString("org.gradle.api.Task"));
+
+    private static final Matcher<ExpressionTree> TASK_GET_PROJECT_METHOD = MethodMatchers.instanceMethod()
+            .onDescendantOf("org.gradle.api.Task")
+            .named("getProject");
+
+    public static final String VIOLATION_MESSAGE = "Don't call `getProject()` in task actions";
+
+    public static final String SUMMARY =
+            """
+        Don't call `getProject()` in task actions. Large, mutable Gradle model types like `Gradle`, `Settings`, or
+        `Project` should not be passed into tasks as inputs. Instead, your tasks should take in the "smallest" type
+        required for the task's functionality. For example, instead of taking in `Project` to later do
+        `project.version`, you should declare the project version as a `Property<String>`.
```

Comment:
Might want to talk about injecting services too (eg using injected `FileOperations` rather than `project#copy`, or `ExecOperations` rather than `project#exec` - basically everything [here](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution) - might want to add a page to the guide (with examples in Java) about this.